### PR TITLE
feat(analytics): extract analytics package and unify event schema

### LINKED ISSUE #4746: feat(analytics): extract analytics package and unify event schema
Part of epic #4338 (Wave 2).

No analytics package exists; 5 files import `FirebaseAnalytics` directly (`app_router.dart`, `error_analytics_tracker.dart`, `feed_performance_tracker.dart`, `screen_analytics_service.dart`, `video_loading_metrics.dart`) and ~9 observability services (~2,951 LOC) live unpackaged in `lib/services/`. The Bloc-error→Crashlytics piece is already done (`DivineBlocObserver` + `lib/observability/`), so this is **analytics-only**.

### Scope
- Extract an analytics package with a unified event schema and a single Firebase wrapper; route screen/widget tracking through it (ideally via a BlocObserver-style hook), not from the UI layer.

### Done when
- UI no longer imports analytics services directly; one event schema; one Firebase surface.

Implementation support: @untreu2 where needed.

---
Audit context: supersedes the analytics half of #3625 (the BlocObserver half already shipped). Supersede only once merged + verified on `main`.
Parent epic: #4338 · Wave 2

### PR DESCRIPTION
**Pull request by untreu2**
## What

Closes #4746 (Wave 2 of epic #4338).

Extracts the unpackaged analytics/observability services from `lib/services/` into a new `mobile/packages/analytics` package with a single Firebase analytics surface and a shared event schema.

## Changes

- **New `analytics` package** (workspace member) owns:
  - the event-sink abstraction (`AnalyticsEventSink` + `NoOpAnalyticsEventSink`),
  - the single Firebase wrapper (`FirebaseAnalyticsEventSink`),
  - the unified schema (`AnalyticsSurface` / `AnalyticsParam` / `SurfaceLoadResult`),
  - the trackers (`ScreenAnalyticsService`, `FeedPerformanceTracker`, `SurfacePerformanceTracker`, `VideoLoadingMetrics`, `ErrorAnalyticsTracker`) and the page-load observer/history.
- **One Firebase surface:** the two trackers that previously bypassed the wrapper (`error_analytics_tracker`, `video_loading_metrics`) now route through `AnalyticsEventSink`; `feed_performance_tracker` migrated onto the sink with a `sink:` test seam matching the other trackers. No app code references `FirebaseAnalytics` directly anymore.
- **Decoupling:** `VideoLoadingMetrics` no longer reaches the app's media-cache global — it takes an injected `CacheMetricsProvider`, wired once in `initializeMediaCache()`.
- Rewired ~30 app importers to the `package:analytics/analytics.dart` barrel and removed the app's now-transitive direct `firebase_analytics` dependency (`pubspec.lock` shows it move from `direct main` → `transitive`).
- Moved analytics-only tests into the package; `page_load_observer`'s widget test stays in the app (needs app localizations).
- Review follow-up: added test seams and package tests for `ErrorAnalyticsTracker` and `VideoLoadingMetrics`, pinned the package `mocktail` dev dependency, and fixed `VideoLoadingMetrics.markLoadingError` so short error messages are recorded instead of throwing during truncation.

## Tests

- 77 package tests pass, including coverage for `ErrorAnalyticsTracker` and `VideoLoadingMetrics`; affected app tests (bloc/comments/page-load-observer/feed/search/video detail widgets) pass in the pre-push changed-file suite.
- `flutter analyze` clean for both the package and the app.

## Notes

- The issue listed `app_router.dart` as a `FirebaseAnalytics` importer; it is not — the real fifth importer was the wrapper itself. Corrected in this PR.
- The package mirrors the app's `analysis_options.yaml` relaxations so the moved code introduces no new lint debt while staying on the `very_good_analysis` baseline.
- Out of scope (called out for a follow-up): the Nostr `analytics_service.dart` (Kind 22236 view events) is unrelated to the Firebase surface and was intentionally left in the app.

### CODE REVIEW

File: mobile/packages/analytics/lib/src/video_loading_metrics.dart [hunk 1]
```diff
@@ -1,22 +1,35 @@
...
 
-  final FirebaseAnalytics _analytics = FirebaseAnalytics.instance;
+  final AnalyticsEventSink _analytics = FirebaseAnalyticsEventSink();
   final Map<String, _VideoLoadingSession> _activeSessions = {};
 
+  /// Optional source of media-cache metrics, wired by the app at startup.
+  ///
+  /// When null, periodic cache-metrics reporting is skipped (keeps the
+  /// package usable without a cache, e.g. in tests).
+  CacheMetricsProvider? cacheMetricsProvider;
```
Comment (realmeylisdev) [inline mobile/packages/analytics/lib/src/video_loading_metrics.dart:38]: **Blocking (per `testing.md` — "Extracted Packages Must Ship with Tests").** This file has no test in the package or anywhere in the repo, yet this PR adds *new* branching logic here: the `CacheMetricsProvider` seam and the `if (provider == null) return;` guard in `reportCacheMetrics()` below, reached via the 50-load cadence in `_maybeReportCacheMetrics()`. Please add a unit test covering at least: (a) `reportCacheMetrics()` no-ops when `cacheMetricsProvider` is null, and (b) it forwards `provider().toMap()` to the sink when wired (mock the sink). Moving this file out of `lib/services/` also drops it from the `untested_services_floor` ratchet with no replacement gate, so this is the moment to add coverage.

File: mobile/packages/analytics/lib/src/error_analytics_tracker.dart [hunk 2]
```diff
@@ -12,9 +13,7 @@ class ErrorAnalyticsTracker {
   factory ErrorAnalyticsTracker() => _instance;
   ErrorAnalyticsTracker._internal();
 
-  // Lazy initialization to avoid Firebase dependency during construction
-  FirebaseAnalytics? _analytics;
-  FirebaseAnalytics get analytics => _analytics ??= FirebaseAnalytics.instance;
+  final AnalyticsEventSink _analytics = FirebaseAnalyticsEventSink();
```
Comment (realmeylisdev) [inline mobile/packages/analytics/lib/src/error_analytics_tracker.dart]: Same extraction-test gap: this tracker ships with no test. The switch to the sink here is also a behavior change worth pinning — the old `analytics` getter threw if `FirebaseAnalytics.instance` was accessed before `Firebase.initializeApp()`, whereas the sink now fails closed to a no-op. A small mock-sink test (assert event name + params are forwarded) would lock the new contract.

File: mobile/packages/analytics/pubspec.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  profile_repository:
+    path: ../profile_repository
+  unified_logger:
+    path: ../unified_logger
+
+dev_dependencies:
+  flutter_test:
+    sdk: flutter
+  very_good_analysis: ^10.2.0
+  mocktail: any
```
Comment (realmeylisdev) [inline mobile/packages/analytics/pubspec.yaml]: Nit: pin `mocktail: ^1.0.4` to match every other package under `mobile/packages/` and `code_style.md` ("always use the latest stable version"). A bare `any` admits a future breaking major into the resolver.

File: mobile/packages/analytics/lib/src/feed_performance_tracker.dart [hunk 4]
```diff
@@ -37,27 +38,22 @@ class FeedPerformanceTracker {
...
-  }
+  FeedPerformanceTracker.testInstance({AnalyticsEventSink? sink})
+    : _analytics = sink ?? const NoOpAnalyticsEventSink();
+
+  final AnalyticsEventSink _analytics;
 
-  static FirebaseAnalytics? _initAnalytics() {
+  /// Resolves the production sink, falling back to a no-op when Firebase is
+  /// not initialized (e.g. during tests that construct the singleton).
+  static AnalyticsEventSink _resolveDefaultSink() {
```
Comment (realmeylisdev) [inline mobile/packages/analytics/lib/src/feed_performance_tracker.dart]: Nit: this `try/catch` is unreachable — `FirebaseAnalyticsEventSink()` is a pure field-store and can't throw (Firebase is resolved lazily inside the sink, which already fails closed). The sibling trackers (`error_analytics_tracker`, `video_loading_metrics`) just use `final _analytics = FirebaseAnalyticsEventSink();` directly; this helper could collapse to `=> FirebaseAnalyticsEventSink();` or be inlined for consistency.

**Review by realmeylisdev — CHANGES_REQUESTED** [review]
## Formal review — analytics package extraction (closes #4746, Wave 2 of #4338)

**Verdict: Changes requested** — one substantive item (the extraction-test rule), plus three non-blocking nits. The extraction itself is clean: I verified it is behavior-preserving and has no backend-contract impact. The request is narrowly about the repo's own "extracted packages ship with tests" rule.

### What I verified (all confirmed)
- **Behavior preserved / improved.** Firebase is still resolved lazily with retry — `FirebaseAnalyticsEventSink` defers `FirebaseAnalytics.instance` to first `logEvent`/`logScreenView` and fails closed when Firebase is uninitialized. `error_analytics_tracker` and `video_loading_metrics`, which previously touched `FirebaseAnalytics.instance` eagerly (throw-on-uninit), now degrade to a no-op — a strict robustness improvement. No event name or parameter key changed in any move.
- **Cache-metrics decoupling is correct.** `VideoLoadingMetrics.cacheMetricsProvider` is wired in `initializeMediaCache()` via a live closure `() => openVineMediaCache.metrics` (fresh read each call). Both `initializeMediaCache` and `reportCacheMetrics` early-return on web, so web is unchanged; the 50-load cadence guarantees init runs first.
- **Baseline ratchets are correct and required.** Removing the 7 `ui_service_imports.txt` entries is mandatory — each file no longer imports any service (verified), so leaving them would fail the STALE check. Removing `error_analytics_tracker`/`video_loading_metrics` from `untested_services.txt` is mandatory — they left `lib/services/`, so leaving them would also fail STALE.
- **Lock is clean.** `firebase_analytics` moves `direct main → transitive` with no version change; the new workspace member correctly has no lock entry (matches every sibling path package).
- **No backend contract impact.** Searched every divine-* backend repo: nothing references the emitted Firebase event names or parameter keys, there is no mobile-analytics ingestion endpoint (events go only to Firebase/Google), and the one backend-relevant path — Nostr **kind 22236**, consumed by divine-funnelcake's relay → ClickHouse → Gorse — is left untouched (`analytics_service.dart` is not in this PR).
- **No orphaned imports / no removed-API breakage.** All importers rewired to the barrel; no remaining `package:openvine/services/<moved>` or `package:firebase_analytics` imports in app `lib`/`test`; the removed public `ErrorAnalyticsTracker.analytics` getter and old `testInstance(analytics:)` seam have no remaining callers.

### Blocking
- **The extracted package ships without tests for two of its trackers.** `error_analytics_tracker` and `video_loading_metrics` have no test in the package (or anywhere in the repo), and the package has no coverage gate. `testing.md` ("New and Extracted Packages Must Ship with Tests… the extraction PR is incomplete without them… cover all public methods") applies directly. This is amplified because `video_loading_metrics` gains **new** branching logic in this PR (the `CacheMetricsProvider` seam + the `if (provider == null) return;` guard) that is uncovered, and because both files leave `lib/services/` — i.e. out from under the `untested_services_floor` ratchet — with no replacement gate. See the inline comments for the specific cases to cover. If the team prefers to treat this as tracked pre-existing debt (both were already on `untested_services.txt`, though without a `# defer:` annotation), that's a maintainer call — but per the written rule this is the place to add at least the new-logic coverage.

### Nits (non-blocking)
- `pubspec.yaml`: `mocktail: any` → pin `^1.0.4` to match every other `mobile/packages/*` and `code_style.md`. (inline)
- `feed_performance_tracker.dart`: dead `try/catch` in `_resolveDefaultSink()`. (inline)
- `feed_performance_tracker`'s own tests don't exercise the new `sink:` seam (they use the no-op default); the surface/screen tracker tests do. Minor.

Well-scoped extraction overall — the only ask is to close the test gap the repo's own rules call for.

**Comment by NotThatKindOfDrLiz** [comment]
Review follow-up pushed in 4da58e12f.\n\nWhat changed:\n- Added test-only analytics sink injection seams to ErrorAnalyticsTracker and VideoLoadingMetrics.\n- Added package tests covering ErrorAnalyticsTracker event names/parameters and error-count behavior.\n- Added package tests covering VideoLoadingMetrics completion/error/cache reporting/no-op behavior.\n- Fixed VideoLoadingMetrics.markLoadingError so short error messages are recorded instead of throwing during truncation.\n- Pinned analytics package mocktail dev dependency to ^1.0.4 instead of any.\n- Updated the PR body test count/details; it still closes #4746.\n\nLocal verification:\n- mise exec flutter@3.44.0 -- flutter test (mobile/packages/analytics): 77 passed.\n- mise exec flutter@3.44.0 -- flutter analyze (mobile/packages/analytics): clean.\n- mise exec flutter@3.44.0 -- flutter analyze (mobile): clean.\n- mise exec flutter@3.44.0 -- flutter test test/services/page_load_observer_test.dart: passed.\n- Pre-push hook: generated files OK and changed-file test suite passed.

**Review by NotThatKindOfDrLiz — APPROVED** [review]
Final pass looks good. The extraction now includes package coverage for the extracted ErrorAnalyticsTracker and VideoLoadingMetrics surfaces, the mocktail dev dependency is pinned, and the short-error-message truncation bug in video loading error analytics is fixed.\n\nVerified locally and on GitHub: package tests, app analyze, page-load observer test, pre-push changed-file suite, linked-issue guard, and all required checks are green on 4da58e12f.

**Comment by NotThatKindOfDrLiz** [comment]
@realmeylisdev I pushed 4da58e12f to address your blocking review item and the dependency nit:\n\n- Added package tests for ErrorAnalyticsTracker.\n- Added package tests for VideoLoadingMetrics, including cache provider wired/no-provider behavior and loading error events.\n- Added test-only sink injection seams for both extracted trackers.\n- Pinned mocktail to ^1.0.4.\n\nAll checks are green on the new head. Please re-review when you have a chance so GitHub can clear the stale CHANGES_REQUESTED decision.

**Review by realmeylisdev — APPROVED** [review]
## Re-review — approved

Thanks @NotThatKindOfDrLiz. I re-verified against `4da58e12f` rather than taking the green checks on faith — the blocking item is fully resolved and the dependency nit is fixed. CI is green on this head (Analyze, Format, Tests, Generated Files, VGV Tag Gate all pass).

**Blocking item — resolved:**
- `VideoLoadingMetrics` and `ErrorAnalyticsTracker` now have package tests via `@visibleForTesting testInstance({sink})` seams. The seams are behavior-neutral in production — `_internal()` still constructs `FirebaseAnalyticsEventSink()`.
- Both cases I specifically called out are covered with real assertions on a recording sink: `reportCacheMetrics()` no-ops when no provider is wired, and forwards the full `video_cache_performance` parameter map (including `hit_rate`) when wired.
- `ErrorAnalyticsTracker` tests cover all eight tracking methods (event name + exact parameter maps), occurrence counting, and reset.
- Bonus: `_truncate(...)` fixes a latent `RangeError` — the old `errorMessage!.substring(0, 100)` threw whenever the message was shorter than 100 chars, and the `markLoadingError(..., 'short')` test now exercises exactly that path. Truncation behavior for messages over 100 chars is unchanged.

**Dependency nit — resolved:** `mocktail` pinned to `^1.0.4`, matching every sibling package.

**Remaining nits (non-blocking, your call):**
- Dead `try/catch` in `FeedPerformanceTracker._resolveDefaultSink()` — the sink constructor can't throw.
- `feed_performance_tracker`'s own tests still use the no-op default rather than the `sink:` seam (the surface/screen tracker tests do assert via the sink).

Neither blocks merge. Clean extraction with no backend-contract impact — nice work on the added coverage.

**Comment by NotThatKindOfDrLiz** [comment]
Closeout update on the current head (938ba422b1):

- Rebased the PR onto current origin/main so mergeability is clean.
- Removed the remaining dead analytics sink fallback helpers. FeedPerformanceTracker, ScreenAnalyticsService, and SurfacePerformanceTracker now construct FirebaseAnalyticsEventSink directly; Firebase still resolves lazily and fails closed inside the sink, so production behavior is unchanged.
- Resolved the stale review threads. The tracker test gaps, mocktail pin, and dead try/catch feedback are all addressed.
- Verified locally with packages/analytics tests, flutter analyze packages/analytics, the repo pre-push analyzer, generated-files check, and changed-file test suite.
- Verified GitHub checks are green: mergeability, Generated Files, Build Flutter Web Preview, semantic PR, Format, Analyze, Tests, VGV Tag Gate, Mobile CI, and CLA.

The PR body links and closes #4746, and that issue has the detailed scope for this extraction.

**Review by NotThatKindOfDrLiz — APPROVED** [review]
Final pass on 938ba422b1: approved. The PR is rebased onto current main, linked to and closing #4746, all review threads are resolved, and all GitHub checks are green. I re-reviewed the analytics package extraction and the final sink-helper cleanup; no remaining blockers found.
